### PR TITLE
Added support for Openstack Image V1 API.

### PR DESCRIPTION
**Pull request by yaso195**
Currently, we are using the Openstack V2 Image API to upload images to
the Openstack. However, some Openstack accounts may not be supporting V2
yet. So for such accounts, V1 will be used to create/upload images.

@mbhinder @zquestz  @lilirui @tw4dl @rjdavidburke

### CODE REVIEW

File: virtualmachine/openstack/util.go [hunk 1]
```diff
@@ -107,43 +109,124 @@ func getBlockStorageClient(vm *VM) (*gophercloud.ServiceClient, error) {
...
+	defer resp.Body.Close()
+
+	body, _ := ioutil.ReadAll(resp.Body)
+	bodyStr := string(body)
+	if resp.StatusCode != 300 {
+		return 0, fmt.Errorf("Image API version request returned bad response, %s", bodyStr)
+	}
+
+	// Prefer V2 over V1
+	if strings.Contains(bodyStr, "\"id\": \"v2.0\"") || strings.Contains(bodyStr, "\"id\": \"v2.1\"") || strings.Contains(bodyStr, "\"id\": \"v2.2\"") {
```
Comment (zquestz) [inline virtualmachine/openstack/util.go]: Can't we match with a regexp and look for v2.x?

**Comment by zquestz** [comment]
This looks very sane to me. I would refactor the string comparisons to use regexps so they aren't so brittle, but other than that LGTM.

File: virtualmachine/openstack/util.go [hunk 2]
```diff
@@ -107,43 +109,124 @@ func getBlockStorageClient(vm *VM) (*gophercloud.ServiceClient, error) {
...
+	// Send the request to upload the image
+	resp, err := versionClient.Do(versionReq)
+	if err != nil {
+		return 0, fmt.Errorf("Failed to send a image API version request")
+	}
+	defer resp.Body.Close()
+
+	body, _ := ioutil.ReadAll(resp.Body)
+	bodyStr := string(body)
+	if resp.StatusCode != 300 {
```
Comment (lilirui) [inline virtualmachine/openstack/util.go]: Should use HTTP codes defined in Golang, https://golang.org/src/net/http/status.go

File: virtualmachine/openstack/util.go (hunk 1 above)
Comment (lilirui) [inline virtualmachine/openstack/util.go]: 👍

File: virtualmachine/openstack/util.go [hunk 3]
```diff
@@ -192,8 +285,10 @@ func uploadImage(tokenID string, imageEndpoint string, imageID string, imagePath
 	}
 	defer resp.Body.Close()
 
-	if resp.StatusCode != 204 {
-		return fmt.Errorf("Upload Image request returned bad response, %s", resp.Body)
+	body, _ := ioutil.ReadAll(resp.Body)
+	if (imageApiVersion == 1 && resp.StatusCode != 200) ||
+		(imageApiVersion == 2 && resp.StatusCode != 204) {
```
Comment (lilirui) [inline virtualmachine/openstack/util.go]: ditto

File: virtualmachine/openstack/util.go [hunk 4]
```diff
@@ -192,8 +285,10 @@ func uploadImage(tokenID string, imageEndpoint string, imageID string, imagePath
...
+		return fmt.Errorf("Upload Image request returned bad response, %s", string(body))
```
Comment (lilirui) [inline virtualmachine/openstack/util.go]: `image` instead of `Image`?

**Comment by lilirui** [comment]
All error messages should be lower case.

File: virtualmachine/openstack/util.go (hunk 2 above)
Comment (yaso195) [inline virtualmachine/openstack/util.go]: Done

File: virtualmachine/openstack/util.go (hunk 3 above)
Comment (yaso195) [inline virtualmachine/openstack/util.go]: Done

File: virtualmachine/openstack/util.go (hunk 4 above)
Comment (yaso195) [inline virtualmachine/openstack/util.go]: Done

File: virtualmachine/openstack/util.go (hunk 1 above)
Comment (yaso195) [inline virtualmachine/openstack/util.go]: Done

**Pull request merged**